File: python/starter/src/t0_provider_starter/template/src/provider/complete_manual_aml_check.py
"""Complete a manual AML check for a payout.

Go equivalent: internal/complete_manual_aml_check.go → CompleteManualAmlCheck()

Pay-Out Provider role — Step 2.6 (optional). If your pay_out handler returned
manual_aml_check instead of accepted (see handler/payment.py), run your AML
check out-of-band and report the outcome with this call. On approval the
response carries updated amounts and quote ids re-approved by the pay-in
provider; on rejection the payout must not proceed.
"""

from __future__ import annotations

import logging

from t0_provider_sdk.api.tzero.v1.payment.network_connect import NetworkServiceClient
from t0_provider_sdk.api.tzero.v1.payment.network_pb2 import CompleteManualAmlCheckRequest

logger = logging.getLogger(__name__)


async def complete_manual_aml_check(
    network_client: NetworkServiceClient,
    payment_id: int,
) -> None:
    try:
        response = await network_client.complete_manual_aml_check(
            CompleteManualAmlCheckRequest(
                payment_id=payment_id,
                approved=CompleteManualAmlCheckRequest.Approved(),
                # If your check failed, report a rejection instead:
                # rejected=CompleteManualAmlCheckRequest.Rejected(reason="AML check failed"),
            ),
        )

        if response.HasField("approved"):
            # Pay-in provider approved the updated quotes — proceed with the payout using
            # the updated amounts, then report the outcome via finalize_payout.
            logger.info(
                "Manual AML check approved for payment %d: pay_out_amount=%s settlement_amount=%s "
                "pay_out_quote_id=%d pay_out_client_quote_id=%s",
                payment_id,
                response.approved.pay_out_amount,
                response.approved.settlement_amount,
                response.approved.pay_out_quote_id,
                response.approved.pay_out_client_quote_id,
            )
        elif response.HasField("rejected"):
            # Pay-in provider rejected the updated quotes — do NOT proceed with the payout.
            logger.info("Updated quotes rejected for payment %d, do not proceed with the payout", payment_id)
    except Exception:
        logger.exception("Error completing manual AML check")

File: java/starter/template/src/main/java/network/t0/provider/internal/CompleteManualAmlCheck.java
package network.t0.provider.internal;

import network.t0.sdk.proto.tzero.v1.payment.CompleteManualAmlCheckRequest;
import network.t0.sdk.proto.tzero.v1.payment.CompleteManualAmlCheckResponse;
import network.t0.sdk.proto.tzero.v1.payment.NetworkServiceGrpc;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Pay-Out Provider role — Step 2.6 (optional).
 *
 * <p>If your payOut handler returned manual_aml_check instead of accepted (see
 * handler/PaymentHandler.java), run your AML check out-of-band and report the
 * outcome with this call. On approval the response carries the updated amounts
 * and quote ids (re-approved by the pay-in provider via the "last look"); on
 * rejection the payout must not proceed.
 */
public class CompleteManualAmlCheck {

    private static final Logger log = LoggerFactory.getLogger(CompleteManualAmlCheck.class);

    public static void complete(
            NetworkServiceGrpc.NetworkServiceBlockingStub networkClient,
            long paymentId) {
        try {
            CompleteManualAmlCheckResponse response = networkClient.completeManualAmlCheck(
                    CompleteManualAmlCheckRequest.newBuilder()
                            .setPaymentId(paymentId)
                            .setApproved(CompleteManualAmlCheckRequest.Approved.newBuilder().build())
                            // If your check failed, report a rejection instead:
                            // .setRejected(CompleteManualAmlCheckRequest.Rejected.newBuilder()
                            //         .setReason("AML check failed").build())
                            .build());

            switch (response.getResultCase()) {
                case APPROVED -> {
                    // Pay-in provider approved the updated quotes — proceed with the payout using
                    // the updated amounts, then report the outcome via finalizePayout.
                    CompleteManualAmlCheckResponse.Approved approved = response.getApproved();
                    log.info("Manual AML check approved for payment {}: payOutAmount={}, settlementAmount={}, quoteId={}, clientQuoteId={}",
                            paymentId, approved.getPayOutAmount(), approved.getSettlementAmount(),
                            approved.getPayOutQuoteId(), approved.getPayOutClientQuoteId());
                }
                case REJECTED ->
                        // Pay-in provider rejected the updated quotes — do NOT proceed with the payout.
                        log.warn("Updated quotes rejected for payment {}, do not proceed with the payout", paymentId);
                default -> log.warn("Unknown result type for payment {}", paymentId);
            }
        } catch (io.grpc.StatusRuntimeException e) {
            log.error("Error completing manual AML check: {} - {}", e.getStatus().getCode(), e.getMessage());
        }
    }
}
